### PR TITLE
HDDS-9544. Incorrect pipeline ID and state for closed container.

### PR DESCRIPTION
**Pull request by aryangupta1998**
## What changes were proposed in this pull request?

For closed containers, the pipeline ID keeps on changing on every request, it's randomly generating everytime we execute the container info command. To address this issue we directly fetch the pipeline ID from scm client via container info and display it in the implementation of container info command instead of displaying the random pipeline ID assigned during the creation of pipeline when container are closed.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9544

## How was this patch tested?

Tested Manually.
Output:
When the container is open,

``` bash
[root@ozone-ozone-1 ~]# ozone admin container info 2003
Container id: 2003
Pipeline id: 063c7d82-dba0-4cf6-95ed-b40fa84196fa
Write PipelineId: 063c7d82-dba0-4cf6-95ed-b40fa84196fa
Write Pipeline State: OPEN
Container State: OPEN
Datanodes: [d57cfdd8-a285-488b-a0ec-7f3477c6bc16/ozone-ozone-1.ozone-ozone.root.hwx.site,
e2da4dbf-fc09-4576-b86c-3bab72d1cfee/ozone-ozone-4.ozone-ozone.root.hwx.site,
dbb43785-b23d-4932-9667-4004bfb8fc2e/ozone-ozone-3.ozone-ozone.root.hwx.site]
Replicas: [State: OPEN; ReplicaIndex: 0; Origin: e2da4dbf-fc09-4576-b86c-3bab72d1cfee; Location: e2da4dbf-fc09-4576-b86c-3bab72d1cfee/ozone-ozone-4.ozone-ozone.root.hwx.site,
State: OPEN; ReplicaIndex: 0; Origin: dbb43785-b23d-4932-9667-4004bfb8fc2e; Location: dbb43785-b23d-4932-9667-4004bfb8fc2e/ozone-ozone-3.ozone-ozone.root.hwx.site,
State: OPEN; ReplicaIndex: 0; Origin: d57cfdd8-a285-488b-a0ec-7f3477c6bc16; Location: d57cfdd8-a285-488b-a0ec-7f3477c6bc16/ozone-ozone-1.ozone-ozone.root.hwx.site]
```

``` bash
[root@ozone-ozone-1 ~]# ozone admin container info 2003
Container id: 2003
Pipeline id: fcb22631-aa1d-4d0b-b15b-58ce6965ea48
Write PipelineId: 063c7d82-dba0-4cf6-95ed-b40fa84196fa
Write Pipeline State: CLOSED
Container State: CLOSING
Datanodes: [e2da4dbf-fc09-4576-b86c-3bab72d1cfee/ozone-ozone-4.ozone-ozone.root.hwx.site,
dbb43785-b23d-4932-9667-4004bfb8fc2e/ozone-ozone-3.ozone-ozone.root.hwx.site,
d57cfdd8-a285-488b-a0ec-7f3477c6bc16/ozone-ozone-1.ozone-ozone.root.hwx.site]
Replicas: [State: OPEN; ReplicaIndex: 0; Origin: e2da4dbf-fc09-4576-b86c-3bab72d1cfee; Location: e2da4dbf-fc09-4576-b86c-3bab72d1cfee/ozone-ozone-4.ozone-ozone.root.hwx.site,
State: OPEN; ReplicaIndex: 0; Origin: dbb43785-b23d-4932-9667-4004bfb8fc2e; Location: dbb43785-b23d-4932-9667-4004bfb8fc2e/ozone-ozone-3.ozone-ozone.root.hwx.site,
State: OPEN; ReplicaIndex: 0; Origin: d57cfdd8-a285-488b-a0ec-7f3477c6bc16; Location: d57cfdd8-a285-488b-a0ec-7f3477c6bc16/ozone-ozone-1.ozone-ozone.root.hwx.site]
```

### CODE REVIEW

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java [hunk 1]
```diff
@@ -105,10 +108,20 @@ public void execute(ScmClient scmClient) throws IOException {
           && spec.root().userObject() instanceof GenericParentCommand
           && ((GenericParentCommand) spec.root().userObject()).isVerbose();
       if (verbose) {
-        LOG.info("Pipeline Info: {}", container.getPipeline());
+        Pipeline pipeline = null;
```
Comment (sadanand48) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java:111]: We can use the same pipeline info object as above (container.getPipeline()) as we have introduced a new field to denote original write pipeline.

**Review by sadanand48 — COMMENTED** [review]
Thanks @aryangupta1998 for the change, Left one comment , also there is a checkstyle failure.

**Comment by errose28** [comment]
Can we update this change to handle [this issue](https://issues.apache.org/jira/browse/HDDS-9544?focusedCommentId=17780526&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17780526) as well? They come from the same method in the code.

**Comment by aryangupta1998** [comment]
> Can we update this change to handle [this issue](https://issues.apache.org/jira/browse/HDDS-9544?focusedCommentId=17780526&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17780526) as well? They come from the same method in the code.

For #createForRead -> #create, we have hardcoded the pipeline state to ALLOCATED in case the containers are closed and we create a new pipeline for read. Anyways, after the pipeline is closed it will be removed from the DB so the state doesn't matter much. As of now, we support 4 states {ALLOCATED, OPEN, DORMANT, CLOSED}, can you suggest which state should we use?

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java [hunk 2]
```diff
@@ -107,8 +109,11 @@ public void execute(ScmClient scmClient) throws IOException {
       if (verbose) {
         LOG.info("Pipeline Info: {}", container.getPipeline());
       } else {
-        LOG.info("Pipeline id: {}", container.getPipeline().getId().getId());
+        LOG.info("Pipeline id: {}",
```
Comment (sadanand48) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java:112]: This also can be reverted as write pipeline Id already maintains the original pipeline

**Comment by sadanand48** [comment]
Current change looks good to me as the original write pipeline info is displayed.

**Comment by sadanand48** [comment]
@errose28 would you like to take a look ?

**Review by sadanand48 — APPROVED** [review]
LGTM

**Comment by errose28** [comment]
> For #createForRead -> #create, we have hardcoded the pipeline state to ALLOCATED in case the containers are closed and we create a new pipeline for read. Anyways, after the pipeline is closed it will be removed from the DB so the state doesn't matter much. As of now, we support 4 states {ALLOCATED, OPEN, DORMANT, CLOSED}, can you suggest which state should we use?

Ideally we would use the actual state of the pipeline if present, or CLOSED if the pipeline is no longer present in SCM.

**Comment by nandakumar131** [comment]
A container is part of only one Pipeline at any point of time. When the container is in `OPEN/CLOSING` state they are managed by Ratis Pipeline.

Once the container is `CLOSED/QUASI_CLOSED`, they are managed using Simple Pipeline.

Simple Pipelines are kinda ephemeral, they are easy to create and are stateless. Simple Pipeline informations are not stored.
 
Whenever a client requests for the Pipeline of `CLOSED/QUASI_CLOSED` Container, Simple Pipeline is created on the fly and handed over to the client, so the PipelineID is randomly generated. This causes the PipelineID to change every time. This is expected behaviour.
 
One improvement that we can do is to try and generate the same PipelineID every time (for the same set of Datanodes) using `UUID.nameUUIDFromBytes` API. By doing this we will be generating the same PipelineID for every call for same set of Datanodes.

**Comment by sadanand48** [comment]
@nandakumar131 , This patch doesn't modify the pipelineInfo/pipelineID that gets printed as part of container info command, but rather adds a new field that stores the RatisPipelineID that the container was initially created on during OPEN/CLOSING states. This can be helpful in debugging issues using the ratisPipelineID say if one has to look into the logs to understand the lifecycle of that pipeline. Using only ephemeral pipeline here is confusing as it is newly generated everytime for a read and one will assume that the pipeline never got created if logs are searched using it.  
Can we still move forward with this patch as a debugging improvement?

**Comment by errose28** [comment]
I agree with @sadanand48 we need to make these troubleshooting CLIs as intuitive as possible. Returning the write pipeline that was used when writing to the container is useful to cross check with history in the logs logs, and if the pipeline is still active it can be cross checked with `ozone admin pipeline list`. The new output specifies that what is being returned is the "write pipeline".

I have never heard of this term "simple pipeline" and I think it is safe to say most Ozone users have not either.  I can't even find evidence to back up this term from the code, because there it defines [simple pipeline as a standalone pipeline](https://github.com/apache/ozone/blob/e360de03375e5204857626b6949e7a2f61721d29/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SimplePipelineProvider.java#L36).

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java [hunk 3]
```diff
@@ -109,6 +111,16 @@ public void execute(ScmClient scmClient) throws IOException {
...
         LOG.info("Pipeline id: {}", container.getPipeline().getId().getId());
       }
+      LOG.info("Write PipelineId: {}",
+          container.getContainerInfo().getPipelineID().getId());
+      try {
+        String pipelineState = scmClient.getPipeline(
+                container.getContainerInfo().getPipelineID().getProtobuf())
+            .getPipelineState().toString();
+        LOG.info("Write Pipeline State: {}", pipelineState);
+      } catch (IOException ioe) {
```
Comment (sadanand48) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java:121]: Check for PipelineNotFoundException specifically

**Comment by nandakumar131** [comment]
>  I have never heard of this term "simple pipeline" and I think it is safe to say most Ozone users have not either.

During the very early stage of Ozone development Standalone Pipeline was created to write data to single datanode (That is where the name 'Standalone' comes from), later it was repurposed to use it in read path. There was a plan to refactor and rename the Standalone Pipeline, that never happened.
Now it's too late for the refactoring. 🙂

@sadanand48  @errose28 
If the change is for making the troubleshooting process easier, I'm +1
The title is misleading since it says 'Incorrect pipeline ID'

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java (hunk 3 above)
Comment (sadanand48) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java:121]:
```suggestion
      } catch (IOException ioe) {
      if (SCMHAUtils.unwrapException(
          ioe) instanceof PipelineNotFoundException) {
        LOG.info("Write Pipeline State: CLOSED");
      } else {
        LOG.error("Failed to retrieve pipeline info");
      }
    }
```

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java [hunk 4]
```diff
@@ -109,6 +111,16 @@ public void execute(ScmClient scmClient) throws IOException {
...
+        LOG.info("Write Pipeline State: CLOSED");
+      }
```
Comment (sadanand48) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java:123]:
```suggestion
      } catch (IOException ioe) {
      if (SCMHAUtils.unwrapException(
          ioe) instanceof PipelineNotFoundException) {
        LOG.info("Write Pipeline State: CLOSED");
      } else {
        LOG.error("Failed to retrieve pipeline info");
      }
    }
```

Comment (aryangupta1998) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java]: Done

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java [hunk 5]
```diff
@@ -109,6 +113,21 @@ public void execute(ScmClient scmClient) throws IOException {
       } else {
         LOG.info("Pipeline id: {}", container.getPipeline().getId().getId());
       }
+      LOG.info("Write PipelineId: {}",
```
Comment (sumitagrawl) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java:116]: Already logging pipeline Id in if/else block, this is redundant, above can be modified if requried.

**Review by sumitagrawl — COMMENTED** [review]
@aryangupta1998 Thanks for working over this, have a minor comment

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java (hunk 5 above)
Comment (aryangupta1998) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java:116]: Both are different, in else we have Pipeline Id(container.getPipeline().getId().getId()) and below we have Write Pipeline Id(container.getContainerInfo().getPipelineID().getId()).

**Comment by sumitagrawl** [comment]
> simple pipeline as a standalone pipeline

@nandakumar131 As checked code, createPipelineForRead is not more used and just returning random pipelineId. What if this code is changed to return existing pipeline Id itself? No need use pipelineProvider to create pipeline for read.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java (hunk 5 above)
Comment (sadanand48) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/InfoSubcommand.java:116]: @sumitagrawl "Write Pipeline ID" represents original pipeline on which the container was created whereas once pipeline is closed we use random pipeline for read. These two values will be the same till pipeline is closed.

**Comment by nandakumar131** [comment]
>  What if this code is changed to return existing pipeline Id itself?

We cannot return the existing PipelineId as it is possible for the Container to get replicated on different datanodes once it is CLOSED. This will cause more confusion.

**Comment by adoroszlai** [comment]
@sumitagrawl please set fix version in Jira when merging PRs

**Pull request merged**